dhcpv6/async.go

```diff
@@ -0,0 +1,214 @@
+	stopping     chan bool
+	receiveQueue chan DHCPv6
+	sendQueue    chan DHCPv6
+	packetsLock  sync.Mutex
+	packets      map[uint32](chan Response)
+}
+
+// NewClientAsync creates an asynchronous client
+func NewClientAsync() *ClientAsync {
+	return &ClientAsync{
+		ReadTimeout:  DefaultReadTimeout,
+		WriteTimeout: DefaultWriteTimeout,
+	}
+}
+
+// OpenForInterface starts the client on the specified interface
+func (c *ClientAsync) OpenForInterface(ifname string, bufferSize int) error {
+	addr, err := GetLinkLocalAddr(ifname)
+	if err != nil {
+		return err
+	}
+	c.LocalAddr = &net.UDPAddr{IP: *addr, Port: DefaultClientPort, Zone: ifname}
+	return c.Open(bufferSize)
+}
+
+// Open starts the client
+func (c *ClientAsync) Open(bufferSize int) error {
+	var (
+		addr *net.UDPAddr
+		ok   bool
+		err  error
+	)
+
+	if addr, ok = c.LocalAddr.(*net.UDPAddr); !ok {
+		return fmt.Errorf("Invalid local address: %v not a net.UDPAddr", c.LocalAddr)
+	}
+
+	// prepare the socket to listen on for replies
+	c.connection, err = net.ListenUDP("udp6", addr)
+	if err != nil {
+		return err
+	}
+	c.stopping = make(chan bool)
+	c.sendQueue = make(chan DHCPv6, bufferSize)
+	c.receiveQueue = make(chan DHCPv6, bufferSize)
+	c.packets = make(map[uint32](chan Response))
+	c.packetsLock = sync.Mutex{}
+
+	var ctx context.Context
+	ctx, c.cancel = context.WithCancel(context.Background())
+	go c.listen(ctx)
+	go c.broadcast(ctx)
+
+	return nil
+}
+
+// Close stops the client
+func (c *ClientAsync) Close() {
+	c.cancel()
+	<-c.stopping
+	<-c.stopping
+	close(c.sendQueue)
+	close(c.receiveQueue)
+	c.connection.Close()
+}
+
+func (c *ClientAsync) listen(ctx context.Context) {
+	defer func() { c.stopping <- true }()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case packet := <-c.receiveQueue:
+			c.receive(packet)
+		}
+	}
+}
+
+func (c *ClientAsync) broadcast(ctx context.Context) {
+	defer func() { c.stopping <- true }()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case packet := <-c.sendQueue:
+			c.send(packet)
+		}
+	}
+}
+
+func (c *ClientAsync) send(packet DHCPv6) {
+	transactionID, _ := GetTransactionID(packet)
+	c.packetsLock.Lock()
+	f := c.packets[transactionID]
+	c.packetsLock.Unlock()
+
+	raddr, err := c.remoteAddr()
+	if err != nil {
+		f <- NewResponse(nil, err)
+		return
+	}
+
+	c.connection.SetWriteDeadline(time.Now().Add(c.WriteTimeout))
+	_, err = c.connection.WriteTo(packet.ToBytes(), raddr)
+	if err != nil {
+		f <- NewResponse(nil, err)
+		return
+	}
+
+	c.receiveQueue <- packet
+}
+
+func (c *ClientAsync) receive(_ DHCPv6) {
+	var (
+		oobdata  = []byte{}
+		received DHCPv6
+	)
+
+	c.connection.SetReadDeadline(time.Now().Add(c.ReadTimeout))
+	for {
+		buffer := make([]byte, maxUDPReceivedPacketSize)
+		n, _, _, _, err := c.connection.ReadMsgUDP(buffer, oobdata)
+		if err != nil {
+			if err, ok := err.(net.Error); ok && err.Timeout() {
+				return
+			}
+			log.Fatalf("Error receiving the message: %s", err)
```

Comment:
don't use `log.Fatalf` in library code - it will log the message and call `os.Exit`, basically ignoring any defer or other cleanup operation, and abruptly terminating the program.

Rather, log the error and return. It would be nice to have an errors channel